nodebuilder/core/constructors.go

```diff
@@ -1,9 +1,112 @@
 package core
 
 import (
-	"github.com/celestiaorg/celestia-node/core"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"net"
+	"os"
+	"path/filepath"
+
+	"go.uber.org/fx"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+
+	"github.com/celestiaorg/celestia-node/libs/utils"
 )
 
-func remote(cfg Config) *core.Client {
-	return core.NewClient(cfg.IP, cfg.Port)
+const xtokenFileName = "xtoken.json"
+
+func grpcClient(lc fx.Lifecycle, cfg Config) (*grpc.ClientConn, error) {
+	var opts []grpc.DialOption
+	if cfg.TLSEnabled {
+		opts = append(opts, grpc.WithTransportCredentials(
+			credentials.NewTLS(&tls.Config{MinVersion: tls.VersionTLS12})),
+		)
+	} else {
+		opts = append(opts, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	}
+	if cfg.XTokenPath != "" {
+		xToken, err := parseTokenPath(cfg.XTokenPath)
+		if err != nil {
+			return nil, err
+		}
+		opts = append(opts, grpc.WithUnaryInterceptor(authInterceptor(xToken)))
+		opts = append(opts, grpc.WithStreamInterceptor(authStreamInterceptor(xToken)))
+	}
+
+	endpoint := net.JoinHostPort(cfg.IP, cfg.Port)
+	conn, err := NewGRPCClient(endpoint, opts...)
+	if err != nil {
+		return nil, err
+	}
+	lc.Append(fx.Hook{
+		OnStart: func(ctx context.Context) error {
```

Comment:
Should we try OnStop as well?

Reply:
I've tried and it shows a lot of error on node stop. Connection closes before we stop fetcher.

Reply:
Thats actually unexpected. Fx closes components in the correct order, i.e. the opposite of starting order, so if GRPCConn is started before Fetcher is started, the conn gonna be stopped after the Fetcher

Reply:
![Uploading Screenshot 2024-12-12 at 02.16.39.png…]()

Reply:
It say "is closing" which means it was stopped and is in progress. 

Which call on Fetcher is giving you this error? We might not need this call. Or we might need to wait until stopping ends in OnStop hook

Reply:
fixed